### PR TITLE
finished day19

### PR DESCRIPTION
**Pull request by KoenvdBerg**

### CODE REVIEW

File: aoc2018/src/main/scala/day19.scala [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    def setr(in: Array[Int]): Array[Int] = {in(c) = in(a); in}
+    def seti(in: Array[Int]): Array[Int] = {in(c) = a; in}
+    def gtir(in: Array[Int]): Array[Int] = {in(c) = if a > in(b) then 1 else 0; in}
+    def gtri(in: Array[Int]): Array[Int] = {in(c) = if in(a) > b then 1 else 0; in}
+    def gtrr(in: Array[Int]): Array[Int] = {in(c) = if in(a) > in(b) then 1 else 0; in}
+    def eqir(in: Array[Int]): Array[Int] = {in(c) = if a == in(b) then 1 else 0; in}
+    def eqri(in: Array[Int]): Array[Int] = {in(c) = if in(a) == b then 1 else 0; in}
+    def eqrr(in: Array[Int]): Array[Int] = {in(c) = if in(a) == in(b) then 1 else 0; in}
+
+    val ops: Map[String, Array[Int] => Array[Int]] = Map(
```
Comment (nmcb) [inline aoc2018/src/main/scala/day19.scala:34]: nice :)

File: aoc2018/src/main/scala/day19.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      "addr" -> addr, "addi" -> addi, "mulr" -> mulr, "muli" -> muli, "banr" -> banr, "bani" -> bani,
+      "borr" -> borr, "bori" -> bori, "setr" -> setr, "seti" -> seti, "gtir" -> gtir, "gtri" -> gtri,
+      "gtrr" -> gtrr, "eqir" -> eqir, "eqri" -> eqri, "eqrr" -> eqrr)
+
+    def run(input: Vector[Int]): Vector[Int] =
+      val executable: Array[Int] => Array[Int] = ops(op)
```
Comment (nmcb) [inline aoc2018/src/main/scala/day19.scala:40]: could make this safer at runtime with `ops.getOrElse(op, sys.error("boom!"))`

File: aoc2018/src/main/scala/day19.scala [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      executable(input.toArray).toVector
+
+
+
+  private val input: (Vector[Operation], Int) =
+
+    def parseOps(s: String): Option[Operation] =
```
Comment (nmcb) [inline aoc2018/src/main/scala/day19.scala:47]: parsing strings reads better with pattern matching, eg.:

```
s match
  case s"$op $i0 $i1 $i2" => Some(Operation(op, i0.toInt, i1.toInt, i2.toInt))
```

File: aoc2018/src/main/scala/day19.scala [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    val lines: Vector[String] = Source
+      .fromResource(s"day${day}.txt")
+      .getLines
+      .toVector
+
+    (lines.flatMap(parseOps), lines.flatMap(parseIp).head)
+
+  def incrementIp(reg: Vector[Int], ip: Int): Vector[Int] =
+    val tmp: Array[Int] = reg.toArray
+    tmp(ip) = tmp(ip) + 1
```
Comment (nmcb) [inline aoc2018/src/main/scala/day19.scala:67]: better to employ `reg.updated(ip, reg(ip) + 1)`

File: aoc2018/src/main/scala/day19.scala [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    tmp.toVector
+
+
+  def runProgram(prg: Vector[Operation], ip: Int, n: Int = 0)(maxIts: Int, reg: Vector[Int]): Vector[Int] =
```
Comment (nmcb) [inline aoc2018/src/main/scala/day19.scala:71]: love the higher-order application :)

File: aoc2018/src/main/scala/day19.scala [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    if n >= maxIts && maxIts != -1 then reg
+    else
+      val instruction: Try[Operation] = Try(prg(reg(ip)))
+      instruction match
+        case Failure(_) => {println("HIT") ; reg}
```
Comment (nmcb) [inline aoc2018/src/main/scala/day19.scala:76]: it's not clear to me under which condition the failure can occur :|

**Review by nmcb — CHANGES_REQUESTED** [review]
minor comments and some explanation required

File: aoc2018/src/main/scala/day19.scala (hunk 3 above)
Comment (KoenvdBerg) [reply aoc2018/src/main/scala/day19.scala:47]: Ah yes I cleaned it up using the pattern match on strings:)

File: aoc2018/src/main/scala/day19.scala (hunk 4 above)
Comment (KoenvdBerg) [reply aoc2018/src/main/scala/day19.scala:67]: Talking about optimization, just this change alone made my code run in 800ms instead of 1300ms. Nice!

**Review by nmcb — COMMENTED** [review]
did resolve most conversations - just the failure one is unclear to me

File: aoc2018/src/main/scala/day19.scala (hunk 6 above)
Comment (KoenvdBerg) [reply aoc2018/src/main/scala/day19.scala]: Hi I've updated the code and provided lots of documentation. Hope that it's a lot clearer now. If not, let me know, then I'll take another look.

I value the feedback a lot btw, so big thanks!!

Comment (nmcb) [reply aoc2018/src/main/scala/day19.scala]: nice koen - very clear writeup :)

**Review by nmcb — APPROVED** [review]

**Pull request merged**